File: analytical.R
account <- read.csv('data/accounts.csv')
district <- read.csv('district_r.csv')
clients <- read.csv('data/clients.csv')
links <- read.csv('data/links.csv')
cards <- read.csv('data/cards.csv')
loan <- read.csv('loan_r.csv')
transactions <- read.csv('data/transactions.csv')
payment_orders <- read.csv('data/payment_orders.csv')
#join account and district
head(account)
head(district)
# include all the rows of two data frames using "left_join" function to join match "account_id": 
#reference: https://stackoverflow.com/questions/1299871/how-to-join-merge-data-frames-inner-outer-left-right

A_D<- merge(x = account, y =district , by = c("district_id" = "id"), all.x = TRUE) 
# left outer join with same district_id
head(A_D)
#A_D <- account %>% left_join(districts,by=c('district_id' = 'id'))
#reference: https://stackoverflow.com/questions/7531868/how-to-rename-a-single-column-in-a-data-frame
names(A_D)[names(A_D) == "date"] <- "open_date" 
names(A_D)[names(A_D) == "id"] <- "account_id"
names(A_D)[names(A_D) == "name"] <- "district_name"
head(A_D)
A_D<-A_D[c("account_id","district_name","open_date","statement_frequency")]
#adjoin <- adjoin[,c(1,3,4,5)]
head(A_D)
ncol(A_D)
nrow(A_D) #4500

#join "links" table to A_D
head(links)
links1<-links %>% group_by(account_id) %>% summarize(num_customers = n()) 
A_D_L <- merge (x=A_D, y=links1, by = c('account_id' = 'account_id'))
head(A_D_L)
nrow(A_D_L) #4500

#join "card" to links
head(cards)
names(cards)[names(cards) == 'type'] <- "card_type"
names(links)[names(links) == 'id'] <- "links_id"
head(links)
x<-cards$link_id
links$card <- ifelse (links$links_id %in%x,"T","F")
y<-links[(links$card == "T"),]
y$cards = 1

A_D_L$credit_cards <- y$cards[(match(A_D_L$account_id, y$account_id))]
A_D_L$credit_cards[is.na(A_D_L$credit_cards)] <- 0
nrow(A_D_L) # 4500
head(A_D_L)

#join loan to A_D_L
head(loan)
nrow(loan) #682

drop<-c("date")
loan = loan[,!(names(loan) %in% drop)]
head(loan)
ADL_L<- merge(x = A_D_L, y =loan , by = c("account_id" = "account_id"), all.x = TRUE)
head(ADL_L)
#DETERMINE T/F:
ADL_L$loan <-ifelse(ADL_L$id %in% NA, 'F','T')
loan <- ADL_L %>% group_by(account_id) %>% distinct() # FILTER AND GROUP BY ACCOUNT ID
names(loan)[names(loan) == 'amount'] <- "loan_amount"
names(loan)[names(loan) == 'payments'] <- "loan_payments"
names(loan)[names(loan) == 'terms'] <- "loan_term"
head(loan)
drop<-c("id")
loan = loan[,!(names(loan) %in% drop)]
head(loan)
nrow(loan)#4500


#### max& min witrhdrawal: 
head(transactions)
W <- transactions %>% filter(type == 'debit') %>% as.data.frame()
#DEFINE min& max withdrawal: amount withdrawn for the account
transactions$min_withdrawal <- min(W$amount)
transactions$max_withdrawal <- max(W$amount)
head(transactions)
trans <- transactions[,c('account_id','max_withdrawal','min_withdrawal')]
head(trans)
nrow(trans) #1056320
W_join <- loan %>%left_join(trans, by = c('account_id'= 'account_id')) %>% distinct()

head(W_join) # done
nrow(W_join)#4500
#----------

####
# cc_payments: Count of credit payments for the account for all cards

credit_pay <- transactions %>% filter(type == "credit") %>% as.data.frame()
#rlang::last_error()
T_C<-credit_pay %>% group_by(account_id) %>% summarize(cc_payments = n())
#head(W_by)
nrow(T_C)#4500
TC_join<- merge(x = W_join,y =T_C, by = c("account_id" = "account_id"), all.x = TRUE)
head(TC_join)
nrow(TC_join) #4500

####
#min/max_balance: amount of balance in the account
transactions$min_balance <- min(W$balance) # define min balance 
transactions$max_balance <- max(W$balance) #define max balance
head(transactions)
trans <-transactions[,c("account_id","max_balance","min_balance")]
analytical <- TC_join %>%left_join(trans, by = c('account_id'= 'account_id')) %>% distinct()
#nrow(analytical) #4500

write.csv(analytical,"analytical_r.csv",row.names = FALSE)





  

File: analytical.py
import os
import pandas as pd
import numpy as np
os.getcwd()
account = pd.read_csv('/data/accounts.csv')
district = pd.read_csv('/district_py.csv')
clients = pd.read_csv('/data/clients.csv')
links = pd.read_csv('/data/links.csv')
cards = pd.read_csv('/data/cards.csv')
loan= pd.read_csv('/loans_py.csv')
payment_orders = pd.read_csv('/data/payment_orders.csv')
transactions = pd.read_csv('/data/transactions.csv')

#join account and district:
print(account.iloc[:5, :11])
print(district.iloc[:5, :11])

#rename column name
district.rename(columns = {'id':'district_id'}, inplace = True)
district.dtypes
merge_AD = pd.merge(account,district, how= "left",on=['district_id'])
print(merge_AD)
merge_AD.dtypes

merge_AD.rename(columns = {'id':'account_id'}, inplace = True)
merge_AD.rename(columns = {'date':'open_date'}, inplace = True)
merge_AD.rename(columns = {'name':'district_name'}, inplace = True)
merge_AD.dtypes #4500

AD=merge_AD[['account_id','district_name','open_date','statement_frequency']]
AD.dtypes #done

# left join "links" table to AD table:
# group by reference: https://www.shanelynn.ie/summarising-aggregation-and-grouping-data-in-python-pandas/
links_by = links.groupby('account_id')['type'].count().reset_index(name = 'num_customers') # works
#links_by.columns = ['account_id','num_customers'] try but failed
links_by.dtypes # they are both int64, expected get

merge_ADL = pd.merge(AD,links_by, how= "left",on=['account_id'])
print(merge_ADL) # 4500rows, 5 cols

#join "cards" to "links" table:
cards.dtypes
#cards.drop('type', inplace=True, axis=1)
#cards.drop('issue_date', inplace=True, axis=1) # drop useless columns
print(cards.iloc[:5, :11])

links.rename(columns = {'id':'link_id'}, inplace = True)
links.dtypes
cards.rename(columns = {'type':'card_type'}, inplace = True)
cards.dtypes
merge_LC = links.merge(cards, how= "left",on=['link_id'])
len(merge_LC.index)#5369

print(merge_LC.iloc[:5, :11])
merge_LC1=merge_LC.groupby('account_id')['card_type'].count().reset_index(name ='credit_cards')
print(merge_LC1.iloc[:5, :11])
len(merge_LC1.index) # 4500

#join LC to ADL --> ADLC
ADLC=merge_ADL.merge(merge_LC1,how= "left",on=['account_id'])
print(ADLC.iloc[:5, :11])
ADLC.dtypes
len(ADLC.index) #4500

#join ADLC to loans:
print(loan)
print(ADLC)
ADLC_=ADLC.merge(loan,how= "left",on=['account_id'])
ADLC_.dtypes
len(ADLC_.index) #4500

ADLC_.rename(columns = {'id':'loan'}, inplace = True)
ADLC_.rename(columns = {'amount':'loan_amount'}, inplace = True)
#ADLC_.rename(columns = {'date':'open_date'}, inplace = True)
ADLC_.rename(columns = {'payments':'loan_payments'}, inplace = True)
ADLC_.rename(columns = {'terms':'loan_terms'}, inplace = True)

ADLC_.dtypes
len(ADLC_.index) #4500 rows

# drop useless columns:
ADLC_.drop('date', inplace=True, axis=1) # drop columns "date"

#loans_melt.to_csv('analytical_py.csv',index=False)
ADLC_.dtypes

#max_withdrawal: first try, failed.
#print(transactions.iloc[:5, :11])
#max_with=transactions[transactions["type"]=='debit']
#print(max_with)
#min_with=transactions[transactions["type"]=='debit']

#second try:
#merge them:
ADLC_W = ADLC.merge(max_with,how= "left")
# max
max_with1=transactions[["account_id","amount","type"]].groupby(["account_id","type"]).max()
max_with2 = max_with1.pivot_table(index=["account_id"],columns="type",values = "amount")
max_with2.drop("credit",inplace=True,axis=1)
max_with2.rename({"debit":"max_withdrawal"},inplace=True,axis=1)
print(max_with2)
min_with1=transactions[["account_id","amount","type"]].groupby(["account_id","type"]).min()
min_with2 = min_with1.pivot_table(index=["account_id"],columns="type",values = "amount")
min_with2.drop("credit",inplace=True,axis=1)
min_with2.rename({"debit":"min_withdrawal"},inplace=True,axis=1,)

# merge min&max together:
ADLC_M = pd.merge(ADLC_,max_with2, how= "left",on=['account_id'])
print(ADLC_M)
ADLC_M = pd.merge(ADLC_M,min_with2, how= "left",on=['account_id'])
print(ADLC_M)
#4500 rows & 14 columns


# join cc_payments:
cc_payments = payment_orders["account_id"].value_counts()
print(cc_payments)
cc_payments=pd.DataFrame(cc_payments)
cc_payments.reset_index(axis=1,inplace=True)
cc_payments.rename({"index":"cc_payment"},axis=1,inplace=True)
#cc_payments2 = cc_payments.reset_index().rename(columns={cc_payments.index.name:'account_id'})
#print(cc_payments2)
#ADLC_CC=ADLC_M.merge(cc_payments,how="left",on=['account_id'])
#print(ADLC_CC) # 4500rows
ADLC_M.to_csv('analytical_py.csv',index=False)

#join max_balance: 
#print(transactions)
#max_b= transactions[["account_id","balance"]].groupby("account_id")
#max().reset_index(inplace=True)
#print(max_b)
#max_b.rename({"balance":"max_balance"},axis=1,inplace=True)
#min_b = transactions[["account_id","balance"]].groupby("account_id").min().reset_index(inplace=True)
#min_b.rename({"balance":"min_balance"},axis=1,inplace=True)

# join